From 0ed75d9521d9109e6d80099665378b0e441159c7 Mon Sep 17 00:00:00 2001
From: Yongxin Liu <yongxin.liu@windriver.com>
Date: Thu, 13 Jun 2024 10:26:27 +0000
Subject: [PATCH 06/24] rdk/snr: fix const netdev->dev_addr issue

Due to kernel commit adeef3e32146a ("net: constify netdev->dev_addr"),
ether_addr_copy(netdev->dev_addr, mac_addr) and memcpy(netdev->dev_addr,
mac, netdev->addr_len) don't work after that.

Backport kernel commit f3956ebb3bf0 ("ethernet: use eth_hw_addr_set()
instead of ether_addr_copy()"), and commit b357d9717be7 ("ice: Only lock
to update netdev dev_addr") to fix this issue.

Upstream-Status: Inappropriate [RDK is released in tarball only]

Signed-off-by: Yongxin Liu <yongxin.liu@windriver.com>
---
 klm/cpk/ice_main.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/klm/cpk/ice_main.c b/klm/cpk/ice_main.c
index fd9c6ef..10ee6ae 100644
--- a/klm/cpk/ice_main.c
+++ b/klm/cpk/ice_main.c
@@ -2246,7 +2246,7 @@ static int ice_cfg_netdev(struct ice_vsi *vsi)
 	if (vsi->type == ICE_VSI_PF) {
 		SET_NETDEV_DEV(netdev, ice_pf_to_dev(pf));
 		ether_addr_copy(mac_addr, vsi->port_info->mac.perm_addr);
-		ether_addr_copy(netdev->dev_addr, mac_addr);
+		eth_hw_addr_set(netdev, mac_addr);
 		ether_addr_copy(netdev->perm_addr, mac_addr);
 	}
 
@@ -4039,6 +4039,7 @@ static int ice_set_mac_address(struct net_device *netdev, void *pi)
 	struct ice_pf *pf = vsi->back;
 	struct ice_hw *hw = &pf->hw;
 	struct sockaddr *addr = pi;
+	u8 old_mac[ETH_ALEN];
 	enum ice_status status;
 	u8 flags = 0;
 	int err = 0;
@@ -4061,6 +4062,12 @@ static int ice_set_mac_address(struct net_device *netdev, void *pi)
 		return -EBUSY;
 	}
 
+	netif_addr_lock_bh(netdev);
+	ether_addr_copy(old_mac, netdev->dev_addr);
+	/* change the netdev's MAC address */
+	eth_hw_addr_set(netdev, mac);
+	netif_addr_unlock_bh(netdev);
+
 	/* Clean up old MAC filter before changing the MAC address */
 	status = ice_fltr_remove_mac(vsi, netdev->dev_addr, ICE_FWD_TO_VSI);
 	if (status) {
@@ -4078,10 +4085,12 @@ err_update_filters:
 	if (err) {
 		netdev_err(netdev, "can't set MAC %pM. filter update failed\n",
 			   mac);
+		netif_addr_lock_bh(netdev);
+		eth_hw_addr_set(netdev, old_mac);
+		netif_addr_unlock_bh(netdev);
 		return err;
 	}
 
-	memcpy(netdev->dev_addr, mac, netdev->addr_len);
 	/* change the netdev's MAC address */
 	netdev_dbg(vsi->netdev, "updated MAC address to %pM\n",
 		   netdev->dev_addr);
-- 
2.39.3

